### PR TITLE
Fix OpenAPI bearer security

### LINKED ISSUE #816: Assigning a singular `Security` schema to a `Middleware` leads to an error.
### Discussed in https://github.com/RobinTail/express-zod-api/discussions/803

<div type='discussions-op-text'>

<sup>Originally posted by **McMerph** February  8, 2023</sup>
I want to specify that a particular middleware uses JWT Bearer Authentication.
So, I added:
```
security: { type: "bearer", format: "JWT" }
```
to `createMiddleware`.
Unfortunately it didn't work -(
Error is:
```
TypeError: methods[security.type] is not a function
    at {{PROJECT_DIR}}/node_modules/express-zod-api/src/open-api-helpers.ts:801:69
    at {{PROJECT_DIR}}/node_modules/express-zod-api/src/logical-container.ts:40:15
    at Array.map (<anonymous>)
    at mapLogicalContainer ({{PROJECT_DIR}}/node_modules/express-zod-api/src/logical-container.ts:37:28)
    at depictSecurity ({{PROJECT_DIR}}/node_modules/express-zod-api/src/open-api-helpers.ts:800:29)
    at onEndpoint ({{PROJECT_DIR}}/node_modules/express-zod-api/src/open-api.ts:120:25)
    at {{PROJECT_DIR}}/node_modules/express-zod-api/src/routing-walker.ts:46:9
    at Array.forEach (<anonymous>)
    at {{PROJECT_DIR}}/node_modules/express-zod-api/src/routing-walker.ts:45:15
    at Array.forEach (<anonymous>)
```
What am I doing wrong?

---
```
security: { type: "bearer" }
```
doesn't work either.

Actually
```
security: { or: [{ type: "bearer", format: "JWT" }] }
```
or
```
security: { and: [{ type: "bearer", format: "JWT" }] }
```
both work but https://editor.swagger.io/ reports structural errors in this case (screenshot provided).
Both the middleware and the endpoint have an empty input (`input: z.object({})`) if that matters.
image
</div>

### PR DESCRIPTION
**Pull request by McMerph**
This relates to discussion https://github.com/RobinTail/express-zod-api/discussions/803
This PR fixes #816 

---
### Steps to reproduce:
- add middleware like this:
```
createMiddleware({
  security: { type: "bearer" },
...
```
to any endpoint
- call `new OpenAPI({...}).getSpecAsYaml()`

### Actual behaviour:
```
TypeError: methods[security.type] is not a function
    at {{PROJECT_DIR}}/node_modules/express-zod-api/src/open-api-helpers.ts:801:69
    at {{PROJECT_DIR}}/node_modules/express-zod-api/src/logical-container.ts:40:15
    at Array.map (<anonymous>)
    at mapLogicalContainer ({{PROJECT_DIR}}/node_modules/express-zod-api/src/logical-container.ts:37:28)
    at depictSecurity ({{PROJECT_DIR}}/node_modules/express-zod-api/src/open-api-helpers.ts:800:29)
    at onEndpoint ({{PROJECT_DIR}}/node_modules/express-zod-api/src/open-api.ts:120:25)
    at {{PROJECT_DIR}}/node_modules/express-zod-api/src/routing-walker.ts:46:9
    at Array.forEach (<anonymous>)
    at {{PROJECT_DIR}}/node_modules/express-zod-api/src/routing-walker.ts:45:15
    at Array.forEach (<anonymous>)
```

### Expected behaviour
OpenApi spec generates without errors

### CODE REVIEW

File: src/logical-container.ts [hunk 1]
```diff
@@ -106,5 +106,8 @@ export const combineContainers = <T>(
       }
     }
   }
+  if (isObject(a) && "and" in a && Array.isArray(a.and) && a.and.length === 0)
+    return { and: [b as T] };
+
   return { and: [a as T, b as T] };
```
Comment (RobinTail) [inline src/logical-container.ts:112]: I think I want to rewrite this in a slightly more generic way in order to cover same case (empty array) for `b`.

File: tests/unit/logical-container.spec.ts [hunk 2]
```diff
@@ -107,5 +107,8 @@ describe("LogicalContainer", () => {
         or: [{ and: [3, 4, 1] }, { and: [3, 4, 2] }],
       });
     });
+    test("should combine LogicalAnd and flat", () => {
+      expect(combineContainers({ and: [] }, 1)).toEqual({ and: [1] });
```
Comment (RobinTail) [inline tests/unit/logical-container.spec.ts:111]: need to test the same thing with reversed order of arguments

File: src/logical-container.ts (hunk 1 above)
Comment (RobinTail) [reply src/logical-container.ts]: done

File: tests/unit/logical-container.spec.ts (hunk 2 above)
Comment (RobinTail) [reply tests/unit/logical-container.spec.ts]: done

File: tests/unit/__snapshots__/open-api.spec.ts.snap [hunk 3]
```diff
@@ -405,6 +405,68 @@ paths:
...
+                      message: Sample error message
+      requestBody:
+        content:
+          application/json:
+            schema:
+              description: PUT /v1/updateSomething request body
+              allOf:
+                - type: object
+                  properties: {}
+                  required: []
```
Comment (RobinTail) [inline tests/unit/__snapshots__/open-api.spec.ts.snap:464]: we also need to fix this, according to your original findings, @McMerph 
`required` must be omitted when it's empty

Comment (RobinTail) [reply tests/unit/__snapshots__/open-api.spec.ts.snap]: fixed

**Review by RobinTail — APPROVED** [review]
Now I'm happy with it :)

**Comment by McMerph** [comment]
Great! Thank you, @RobinTail !

**Pull request merged**